src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher.cc

```diff
@@ -0,0 +1,213 @@
+
+// Possible to have the server sending PUBLISH with same packet identifier as client PUBLISH before
+// it sends PUBACK, causing server PUBLISH to be put into response deque instead of request deque.
+// TODO: Reverse logic to match requests that have erroneously been put into response deque
+
+MatchKey getMatchKey(mqtt::Message& frame) {
+  return (frame.control_packet_type << 4) | static_cast<uint8_t>(frame.dup) << 3 |
+         (frame.header_fields["qos"] & 0x3) << 1 | static_cast<uint8_t>(frame.retain);
+}
+
+RecordsWithErrorCount<Record> StitchFrames(
+    absl::flat_hash_map<packet_id_t, std::deque<Message>>* req_frames,
+    absl::flat_hash_map<packet_id_t, std::deque<Message>>* resp_frames, mqtt::StateWrapper* state) {
+  std::vector<Record> entries;
+  int error_count = 0;
+
+  // iterate through all deques of requests associated with a specific streamID and find the
+  // matching response
+  for (auto& [packet_id, req_deque] : *req_frames) {
+    // goal is to match the request to the closest appropriate response to the specific control type
+    // based on timestamp
+
+    // get the response deque corresponding to the packet ID of the request deque
+    auto pos = resp_frames->find(packet_id);
+    // note that not finding a corresponding response deque is not indicative of error, as in
+    // case of MQTT packets that do not have responses like Publish with QOS 0
+    std::deque<mqtt::Message> empty_deque;
+    std::deque<mqtt::Message>& resp_deque = (pos != resp_frames->end()) ? pos->second : empty_deque;
+
+    // track the latest response timestamp to compare against request frame's timestamp later.
+    uint64_t latest_resp_ts = resp_deque.empty() ? 0 : resp_deque.back().timestamp_ns;
+    // finding the closest appropriate response from response deque in terms of timestamp and type
+    // for each request in the request deque
+    for (mqtt::Message& req_frame : req_deque) {
+      const MqttControlPacketType control_packet_type =
+          magic_enum::enum_cast<MqttControlPacketType>(req_frame.control_packet_type).value();
+      // If the frame is PUBLISH, and there are duplicates in the deque, then mark the frame as
+      // consumed and match the latest duplicate with its response (if the response exists in the
+      // response deque)
+      if (control_packet_type == MqttControlPacketType::PUBLISH) {
+        std::tuple<uint32_t, uint32_t> unique_publish_identifier = std::tuple<uint32_t, uint32_t>(
+            req_frame.header_fields["packet_identifier"], req_frame.header_fields["qos"]);
+        if (req_frame.type == message_type_t::kRequest &&
+            state->send[unique_publish_identifier] > 0) {
+          state->send[unique_publish_identifier] -= 1;
+          req_frame.consumed = true;
+          continue;
+        }
+
+        if (req_frame.type == message_type_t::kResponse &&
+            state->recv[unique_publish_identifier] > 0) {
+          state->recv[unique_publish_identifier] -= 1;
```

Comment:
Related to my earlier comment about avoiding excess inserts into `state->{recv,send}`. Can we use `::find()` instead of `::operator[]`?